### PR TITLE
[Develop branch] Feature/data migration from existing database to csv

### PR DESCRIPTION
**Pull request by ilko-adamov**

### CODE REVIEW

**Review by apricot13 — CHANGES_REQUESTED** [review]
* Can you rename the import script to 'import_services'?
* Could you also add some documentation into the README for developers? 


* With this test data I get the following error, could you make the error more verbose if the service title is missing? 
image

```
import_id,import_id_reference,name,description,organisation,url,approved,visible_from,visible_to,visible,needs_referral,min_age,max_age,notes,service_taxonomies,contact_name,contact_title,contact_visible,contact_email,contact_phone,location_name,location_latitude,location_longitude,location_address_1,location_city,location_postcode,location_visible,mask_exact_address,preferred_for_post,location_accessibilities,free,cost_option,cost_amount,cost_type,temporarily_closed,schedules_opens_at,schedules_closes_at,scheduled_weekday,links_label,links_url,labels,suitabilities,is_local_offer,send_needs_support,support_description,recent_send_report,outcomes,recent_send_training,parental_involvement,information_sharing,environment_accessibility,how_to_start,future_plans
1,,CSV creation service,"We create blank CSV's showing the structure of the data to be imported.

The description allows for multiple lines. 

NB this field will be shortened to 30 words in some places",TPX Impact,http://www.tpximpact.com,TRUE,2022-05-17,2023-05-17,FALSE,TRUE,0,100,if you wanted to import existing notes on the service this is the place to do so - we can add extra fields for multiple notes if required,semicolon delimited values; please,public contact name,public contacts role,TRUE,email@email.com,92939293,Venue or building name,51.50101,-0.141563,Street address eg Buckingham Palace,Town or area eg london,SW1A 1AA,TRUE,FALSE,FALSE,list; of accessible features (see documentation for a default set),FALSE,Who is this fee for?,1,per half day,,hh:mm:ss,hh:mm:ss,monday,Facebook,http://www.google.com,internal label,comma; delimited; list,TRUE,Autism;Hearing impairment;Visual impairment;Mobility;Cognitive,dfsdfs,,,,,,,,
2,1,,,,,,,,,,,,I am an additional note for the service with import_id 1,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,
3,1,,,,,,,,,,,,,,private contact for service_id 1,privat contacts job role,FALSE,email2@email.com,234234,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,
4,1,,,,,,,,,,,,,,,,,,,Second Venue or building name,51.50101,-0.141563,Street address eg Buckingham Palace,Town or area eg london,SW1A 1AA,TRUE,FALSE,FALSE,Accessible toilet facilities; Hoist,,,,,,,,,,,,,,,,,,,,,,,
5,1,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,OAP,0,per session,,,,,,,,,,,,,,,,,,,
6,1,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,hh:mm:ss,hh:mm:ss,tuesday,,,,,,,,,,,,,,,
7,1,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,Twitter,http://www.google.com,,,,,,,,,,,,,
8,,Second example service,description,Not TPX Impact,http://www.google.com,TRUE,,,TRUE,FALSE,,,,hello; new; data,contact 1,chief biscuit sampler,TRUE,email@emails.com,34234,McVities Head Office,51.50101,-0.141563,McVities Head Office,Biscuitville,BV12 3MV,TRUE,FALSE,FALSE,Accessible toilet facilities;Hearing loop;Car parking;Changing facilities;British sign language (bsl);Disabled car parking space;Hoist;Baby changing facilities;Parking nearby;Bus stop nearby;Building has lift;Wheelchair accessible entrance;Partial wheelchair access,TRUE,,,,,,,,,,,Autism;Learning difficulties;Mental health/acquired brain injury;Visual and / or hearing impediment;Physical disabilities;Older people;Dementia,FALSE,,,,,,,,,,
9,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,
10,,,,TPX Impact,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,
11,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,
12,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,
13,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,, 
```

* I also get the same error when uploading this data:
```
import_id,import_id_reference,name,description,organisation,url,approved,visible_from,visible_to,visible,needs_referral,min_age,max_age,notes,service_taxonomies,contact_name,contact_title,contact_visible,contact_email,contact_phone,location_name,location_latitude,location_longitude,location_address_1,location_city,location_postcode,location_visible,mask_exact_address,preferred_for_post,location_accessibilities,free,cost_option,cost_amount,cost_type,temporarily_closed,schedules_opens_at,schedules_closes_at,scheduled_weekday,links_label,links_url,labels,suitabilities,is_local_offer,send_needs_support,support_description,recent_send_report,outcomes,recent_send_training,parental_involvement,information_sharing,environment_accessibility,how_to_start,future_plans
1,,CSV creation service,"We create blank CSV's showing the structure of the data to be imported.

The description allows for multiple lines. 

NB this field will be shortened to 30 words in some places",TPX Impact,http://www.tpximpact.com,TRUE,2022-05-17,2023-05-17,FALSE,TRUE,0,100,if you wanted to import existing notes on the service this is the place to do so - we can add extra fields for multiple notes if required,semicolon delimited values; please,public contact name,public contacts role,TRUE,email@email.com,92939293,Venue or building name,51.50101,-0.141563,Street address eg Buckingham Palace,Town or area eg london,SW1A 1AA,TRUE,FALSE,FALSE,list; of accessible features (see documentation for a default set),FALSE,Who is this fee for?,1,per half day,,hh:mm:ss,hh:mm:ss,monday,Facebook,http://www.google.com,internal label,comma; delimited; list,TRUE,Autism;Hearing impairment;Visual impairment;Mobility;Cognitive,dfsdfs,,,,,,,,
2,1,,,,,,,,,,,,I am an additional note for the service with import_id 1,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,
3,1,,,,,,,,,,,,,,private contact for service_id 1,privat contacts job role,FALSE,email2@email.com,234234,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,
4,1,,,,,,,,,,,,,,,,,,,Second Venue or building name,51.50101,-0.141563,Street address eg Buckingham Palace,Town or area eg london,SW1A 1AA,TRUE,FALSE,FALSE,Accessible toilet facilities; Hoist,,,,,,,,,,,,,,,,,,,,,,,
5,1,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,OAP,0,per session,,,,,,,,,,,,,,,,,,,
6,1,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,hh:mm:ss,hh:mm:ss,tuesday,,,,,,,,,,,,,,,
7,1,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,Twitter,http://www.google.com,,,,,,,,,,,,,
8,,Second example service,description,Not TPX Impact,http://www.google.com,TRUE,,,TRUE,FALSE,,,,hello; new; data,contact 1,chief biscuit sampler,TRUE,email@emails.com,34234,McVities Head Office,51.50101,-0.141563,McVities Head Office,Biscuitville,BV12 3MV,TRUE,FALSE,FALSE,Accessible toilet facilities;Hearing loop;Car parking;Changing facilities;British sign language (bsl);Disabled car parking space;Hoist;Baby changing facilities;Parking nearby;Bus stop nearby;Building has lift;Wheelchair accessible entrance;Partial wheelchair access,TRUE,,,,,,,,,,,Autism;Learning difficulties;Mental health/acquired brain injury;Visual and / or hearing impediment;Physical disabilities;Older people;Dementia,FALSE,,,,,,,,,,
```

* this should work since the fields without a service name are extra fields with data to be added to services created beforehand.


* in fact I cant get the import to run at all without getting this error. It doesn't tell me which field is blank. 

* You'll also need to check that `import_id` field doesn't contain duplicates either. Since that could mean data added later could be appended to the wrong service. 

* this is a quick outline for some tests that would need to pass for this
`import_services_spec.rb`

```ruby
require "rails_helper"
Rails.application.load_tasks

describe "import services" do
  after(:each) do
    Rake::Task["import_services"].reenable
  end

  describe 'importing services' do
    
    # import_id

    it 'Should fail when import_id is missing from a row' do
    end

    it 'Should fail when import_id is duplicated' do
    end

    it 'Should fail when import_id is not numeric' do
    end

    # name

    it 'Should fail when service name is duplicated' do
    end

    it 'Should fail when service name is empty and import_id_reference is not set' do
    end

    # organisation

    it 'Should create an organisation if one with that name does not exist' do 
    end

    it 'Should create an unnamed organisation if organisation field is blank' do
    end


   # visible

   it 'Should set visible to FALSE when `visible_from` and `visible_to` are set and valid' do
   end


   # nested

   it 'should import multiple contacts if import_id_reference is set' do 
   end

   it 'should import multiple locations if import_id_reference is set' do 
   end

   it 'should import multiple costs if import_id_reference is set' do 
   end

   it 'should import multiple schedules if import_id_reference is set' do 
   end

   it 'should import multiple links if import_id_reference is set' do 
   end


   # custom fields

   it 'should import data for custom fields if custom field already exists if not fail' do 
   end


   # validation

   it 'Fields should match the data format specified in readme' do 
   end

   it 'Should add additional taxonomies if they dont already exist' do 
   end


   it 'Should add additional accessibilities if they dont already exist' do 
   end


   it 'Should add additional labels if they dont already exist' do 
   end


   it 'Should add additional suitabilities if they dont already exist' do 
   end


   it 'Should add additional send_needs_support if they dont already exist' do 
   end





  end
end
```

**Review by em-cd — CHANGES_REQUESTED** [review]
This looks good on the whole @ilko-adamov, thank you!

I've left a couple of comments inline for you to take a look at.

File: app/models/organisation.rb [hunk 1]
```diff
@@ -82,9 +82,9 @@ def self.options_for_services
       tsearch: { prefix: true }
     }
 
-  validates_uniqueness_of :name, if: -> { name.present? }
+  #validates_uniqueness_of :name, if: -> { name.present? }
```
[the remark this answers is not in the document]
Comment (apricot13) [reply app/models/organisation.rb:85]: if they have duplicate org names in the import then the import should fail and allow them to go back and fix it and restart.

**Comment by em-cd** [comment]
Superseded by #245 and #244

**Pull request closed without merging**